core/queries.py

```diff
@@ -0,0 +1,276 @@
+    propagate_trh = (
+        session.query(
+            BatchClass.id.label("batch_id"),
+            func.avg(trh_prop_sub.c.temperature).label("avg_temp"),
+            func.avg(trh_prop_sub.c.humidity).label("avg_rh"),
+            func.avg(trh_prop_sub.c.vpd).label("avg_vpd"),
+        )
+        .outerjoin(propagate_events, propagate_events.c.batch_id == BatchClass.id)
+        .outerjoin(transfer_events, transfer_events.c.batch_id == BatchClass.id)
+        .outerjoin(
+            trh_prop_sub,
+            and_(
+                propagate_events.c.event_time < trh_prop_sub.c.timestamp,
+                trh_prop_sub.c.timestamp < transfer_events.c.event_time,
+            ),
+        )
+        .group_by(BatchClass.id)
+    ).subquery("propagate_trh")
+
+    query = (
+        session.query(
+            BatchClass.id.label("batch_id"),
+            CropTypeClass.name.label("crop_type_name"),
+            BatchClass.tray_size,
+            BatchClass.number_of_trays,
+            weigh_events.c.event_time.label("weigh_time"),
+            propagate_events.c.event_time.label("propagate_time"),
+            transfer_events.c.event_time.label("transfer_time"),
+            harvest_events.c.event_time.label("harvest_time"),
+            LocationClass.zone,
+            LocationClass.aisle,
+            LocationClass.column,
+            LocationClass.shelf,
+            HarvestClass.crop_yield,
+            HarvestClass.waste_disease,
+            HarvestClass.waste_defect,
+            HarvestClass.over_production,
+            (harvest_events.c.event_time - transfer_events.c.event_time).label(
+                "grow_time"
+            ),
+            grow_trh.c.avg_temp.label("avg_grow_temperature"),
+            grow_trh.c.avg_rh.label("avg_grow_humidity"),
+            grow_trh.c.avg_vpd.label("avg_grow_vpd"),
+            propagate_trh.c.avg_temp.label("avg_propagate_temperature"),
+            propagate_trh.c.avg_rh.label("avg_propagate_humidity"),
+            propagate_trh.c.avg_vpd.label("avg_propagate_vpd"),
+        )
+        .join(CropTypeClass, CropTypeClass.id == BatchClass.crop_type_id)
+        # We inner join on weigh_events, because if the batch doesn't have a weigh event
+        # it doesn't really exist, but outer join on the others since they are optional.
+        .join(weigh_events, weigh_events.c.batch_id == BatchClass.id)
+        .outerjoin(propagate_events, propagate_events.c.batch_id == BatchClass.id)
+        .outerjoin(transfer_events, transfer_events.c.batch_id == BatchClass.id)
+        .outerjoin(harvest_events, harvest_events.c.batch_id == BatchClass.id)
+        .outerjoin(LocationClass, LocationClass.id == transfer_events.c.location_id)
+        .outerjoin(HarvestClass, HarvestClass.batch_event_id == harvest_events.c.id)
+        .outerjoin(grow_trh, grow_trh.c.batch_id == BatchClass.id)
+        .outerjoin(propagate_trh, propagate_trh.c.batch_id == BatchClass.id)
+    )
+    return query
```

Comment:
That's quite a query! :).   I like the idea of having this core module containing queries though - feels like we can migrate some other stuff there at some point...

Reply:
Agreed:
https://github.com/alan-turing-institute/CROP/issues/371
https://github.com/alan-turing-institute/CROP/issues/374